Cross origin opener policy reporting

camillelamy opened the pull request:
Adds the notion of reporting and report-only mode to cross-origin opener policy.

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Firefox
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/tree/master/html/cross-origin-opener-policy/reporting
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: [https://bugs.chromium.org/p/chromium/issues/detail?id=1059303](https://bugs.chromium.org/p/chromium/issues/detail?id=1059303)
   * Firefox: …
   * Safari: …


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5518/browsers.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">/browsers.html</a>  ( <a href="https://whatpr.org/html/5518/895fd80...c8265a7/browsers.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">diff</a> )
<a href="https://whatpr.org/html/5518/browsing-the-web.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">/browsing-the-web.html</a>  ( <a href="https://whatpr.org/html/5518/895fd80...c8265a7/browsing-the-web.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">diff</a> )
<a href="https://whatpr.org/html/5518/dom.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">/dom.html</a>  ( <a href="https://whatpr.org/html/5518/895fd80...c8265a7/dom.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">diff</a> )
<a href="https://whatpr.org/html/5518/iana.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">/iana.html</a>  ( <a href="https://whatpr.org/html/5518/895fd80...c8265a7/iana.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">diff</a> )
<a href="https://whatpr.org/html/5518/index.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">/index.html</a>  ( <a href="https://whatpr.org/html/5518/895fd80...c8265a7/index.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">diff</a> )
<a href="https://whatpr.org/html/5518/origin.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">/origin.html</a>  ( <a href="https://whatpr.org/html/5518/895fd80...c8265a7/origin.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">diff</a> )
<a href="https://whatpr.org/html/5518/window-object.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">/window-object.html</a>  ( <a href="https://whatpr.org/html/5518/895fd80...c8265a7/window-object.html" title="Last updated on Nov 19, 2020, 9:30 PM UTC (c8265a7)">diff</a> )

camillelamy commented:
@annevk and @domenic: could you take a look at this spec PR? This is a tentative to spec the first parts of COOP reporting policy. I have been working on an explainer that details what we have in mind (see https://github.com/camillelamy/explainers/blob/master/coop_reporting.md). This spec PR adds:
- the header definition
- the modification of the COOP definition to include reporting
- the reporting of browsing context group switches

What it doesn't add is:
- the reporting of blocked accesses to/from the COOP page

This is because we are still discussing the feasibility of reporting the accesses in enforcement mode. We're reasonably convinced we have a way forward in report-only mode though, so I can add report of the accesses from the page in report only mode to this PR if you want (I though it looked a bit weird to have it in report-only mode only though, so this is why I didn't add them initially).

domenic commented:
At a quick glance this looks reasonable, but it is a bit hard to review since it also contains most of the commits of #5334. I can do a more in-depth review after #5334 lands, and ideally also #5454 since that contains some reporting stuff.

camillelamy commented:
@domenic Thanks! Now that COOP has been merged, this should be more readable.

domenic commented:
Great! Would you mind rebasing one more time, on top of the recently-merged COEP work?

camillelamy commented:
@domenic I have rebased on the COEP work. Note that this PR does not cover the access reporting. Do you want me to put this in this PR or do a follow-up?

domenic commented:
I'm OK either way; please do whichever is more convenient for you. (I can see separate PRs being easier to focus on and being able to land more independently, but also can see the dependency causing problems.)

domenic commented on source:
```diff
@@ -80126,50 +80316,137 @@ interface <dfn>BarProp</dfn> {
+  <var>browsingContext</var>, a <span>sandboxing flag set</span> <var>sandboxFlags</var>, a
+  <span>cross-origin opener policy</span> <var>navigationCOOP</var> and a <span
+  data-x="coop-enforcement-result">cross-origin opener policy enforcement result</span>
+  <var>COOPEnforcementResult</var>:</p>
```
COOPEnforcementResult should be coopEnforcementResult, throughout

domenic commented on source:
```diff
@@ -80126,50 +80316,137 @@ interface <dfn>BarProp</dfn> {
+    data-x="">null</code>, then:</p>
+
+    <ol>
+     <li><p>Let <var>initialNavigationURL</var> be a new empty <span>URL</span>.</p></li>
```
There's no such thing as an empty URL. You could use null, I suppose, but then there needs to be some better handling for the case where the condition on the next line is false and the other algorithms receive null.

domenic commented on source:
```diff
@@ -80193,6 +80470,90 @@ interface <dfn>BarProp</dfn> {
+  data-x="cross-origin opener policy value">cross-origin opener policy value</span>
+  <var>COOPValue</var>, a string <var>disposition</var>, a <span>URL</span>
+  <var>navigationURL</var>, a <span>URL</span> <var>currentDocumentURL</var>, and a
+  <span data-x="concept-request-referrer">referrer</span> <var>referrer</var>:<p>
```
This should use a structure similar to https://html.spec.whatwg.org/#queue-a-cross-origin-embedder-policy-inheritance-violation, in terms of serializing the response URL for reporting if possible, and using a table for the body.

domenic commented on source:
```diff
@@ -80193,6 +80470,90 @@ interface <dfn>BarProp</dfn> {
+  switch when navigating away from a COOP page</dfn> given a string <var>endpoint</var>, a
+  <span>cross-origin opener policy value</span> <var>COOPValue</var>, a string
+  <var>disposition</var>, a <span>URL</span> <var>navigationURL</var>, a <span>URL</span>
+  <var>currentDocumentURL</var>, and a <span>URL</span> <var>initialNavigationURL</var>:<p>
```
This algorithm is very similar to the previous one. Can they be deduplicated or parametrized?

camillelamy replied:
Done. This adds a bit of duplication of the computation of the initial navigation URL in the two call sites of this particular algorithm.

domenic commented on source:
```diff
@@ -80126,50 +80316,137 @@ interface <dfn>BarProp</dfn> {
-    </ul>
+   <li>
+    <p>If <var>browsingContext</var>'s <span>browsing context group</span>'s <span>browsing
+    context set</span>'s <span data-x="list size">size</span> is strictly greater than <code
```
I meant that there is no difference in spec-language between "strictly greater than" and "greater than" (and it's confusing to use the former). There is a difference between "greater than" vs. "greater than or equal to", but people generally don't mistake the former for the latter.

domenic commented:
I'll work on rebasing this on top of #5736 (navigation params).

domenic commented:
OK, rebased. I'd love it if @annevk could review at this point, at least the navigation parts.

One issue is that this might make https://github.com/whatwg/html/issues/1130 worse, by using the "source browsing context" at the end of navigation, as part of [obtain a browsing context to use for a navigation response](https://whatpr.org/html/5518/origin.html#obtain-browsing-context-navigation). (I.e., as part of determining whether a BCG switch is needed.) However, I think in this case it might be justifiable, since it seems like the algorithm legitimately wants to deal with browsing contexts, not with documents or any other information that could be saved off synchronously. Maybe the trick then would be to name this something like "COOP enforcement reporting source browsing context" or something to make it clear it should only be used that way...

camillelamy commented:
If we go on with the changes to redirects in COOP, this will change a bit. In particular, I should be able to remove the dependency on source at the end of the navigation.

camillelamy commented:
@domenic and @annevk: PTAL. I have reworked the CL on top of the changes to the redirect handling, and added the reporting of accesses. This should now cover everything discussed in the explainer.

ArthurSonzogni commented on source:
```diff
@@ -80475,6 +80776,623 @@ interface <dfn>BarProp</dfn> {
+       <var>initialWindowURL</var>, null otherwise.</td>
+      </tr>
+      <tr>
+       <td>source-file</td>
```
1. The spec here is currently using: source-file, line-number, colum-number
2. The [explainer](https://github.com/camillelamy/explainers/blob/master/coop_reporting.md#user-content-emit-reports:~:text=source%2Dfile%2C%20lineno%2C%20colno%3A%20if%20the%20user%20agent%20is%20currently%20executing%20script%20and%20can%20extract%20a%20source%20file's%20URL%2C%20line%20number%20and%20column%20number%20from%20the%20global%20object%2C%20set%20those%20accordingly.to%20a%20window%20it%20opened%2C%20it%20should%20generate%20a%20report%20for%20the%20COOP%20document%20URL%2C%20the%20current%20environment%20and%20the%20following%20body%3A) is using: source-file, lineno, colno
3. The [implementation](https://source.chromium.org/chromium/chromium/src/+/master:content/browser/net/reporting_service_proxy.cc;l=111?q=queuefeaturepolicyviolationreport%20net%2F) of {DocumentPolicy, FeaturePolicy, Deprecation, Intervention, ...}ViolationReports are using sourceFile, lineNumber, columnNumber.

It would be nice to be consistent with (3) I think.

I am guessing you chose (1), because it matches the CSP behavior. This one seems to be the exception instead of the rule. Moreover, every reports inside ReportingObserver are using (3), including CSP.

In the Chrome [implementation](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/frame/csp/csp_violation_report_body.h;l=15;drc=82904949df9a42b8212b7c911e18055341876d01;bpv=1;bpt=1), CSPViolationReportBody derive from LocationReportBody, which is the base class for every reports. As a result, they are all using (3).

Maybe we should also switch to camelCase for every attributes. WDYT?

camillelamy replied:
I see. I tried to be consistent with the CSP spec I found, but if other APIs are using a different convention we can certainly switch. Overall, I would like COOP reporting to be consistent with other reporting APIs.

camillelamy replied:
I have modified the spec PR to use camelCase for the report body names.

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+       <td><var>coopValue</var></td>
+      </tr>
+      <tr>
+       <td>previousResponseUrl</td>
```
This should be previousResponseURL, per https://w3ctag.github.io/design-principles/#casing-rules.

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+      </tr>
+      <tr>
+       <td>previousResponseUrl</td>
+       <td>If <var>coopURL</var> and <var>previousResponseURL</var> are <span>same origin</span> this
```
Let's pull this out into a dedicated step. Then you can properly get each URL's [origin](https://url.spec.whatwg.org/#concept-url-origin), since the "same origin" algorithm only applies to origins, not URLs.

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+      </tr>
+      <tr>
+       <td>referrer</td>
+       <td><var>referrer</var>.</td>
```
```suggestion
       <td><var>referrer</var></td>
```

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+      </tr>
+      <tr>
+       <td>violationType</td>
+       <td>navigation-to-response</td>
```
```suggestion
       <td>"<code data-x="">navigation-to-response</code>"</td>
```

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+   for <var>endpoint</var> with <var>coopURL</var>.</p></li>
+  </ol>
+
+  <p>To <dfn data-x="coop-violation-navigation-from">queue a violation report for browsing context group
```
Similar small fixes are needed for all the other queue algorithms.

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+       <td><var>disposition</var></td>
+      </tr>
+      <tr>
+       <td>effective-policy</td>
```
This should become effectivePolicy, right?

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+      </tr>
+      <tr>
+       <td>sourceFile</td>
+       <td>Extract it from <var>globalObject</var></td>
```
This information isn't extractable from the global object; it's in fact very local information.

At the same time, we have no precise way of talking about this. So I'd add a step 1 which is something like "Let _sourceFIle_, _lineNumber_, and _columnNumber_ be the relevant script URL and problematic position which triggered this report." Compare to similarly vague language at https://html.spec.whatwg.org/#report-the-exception.

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+  <var>coopValue</var>, three <span data-x="URL">URLs</span> <var>coopURL</var>,
+  <var>openedWindowURL</var> and <var>initialWindowURL</var>, three <span
+  data-x="origin">origins</span> <var>coopOrigin</var>, <var>openedWindowOrigin</var>, and
+  <var>openerInitialOrigin</var>,a string <var>propertyName</var>, a <span>global
```
```suggestion
  <var>openerInitialOrigin</var>, a string <var>propertyName</var>, a <span>global
```

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+       <td><var>referrer</var>.</td>
+      </tr>
+      <tr>
+       <td>violationType</td>
```
To align with COEP, I think this should just be "type".

camillelamy commented:
@domenic @annevk I have picked this up again after a few months of being unable to make progress. I have a new rebased version that integrates @domenic's comments. PTAL.

domenic commented on source:
```diff
@@ -79917,6 +79946,11 @@ interface <dfn>BarProp</dfn> {
+   <li><p><span data-x="coop-check-access-report">Check if an access between two browsing contexts
+   should be reported</span>, given the <span>current global object</span>'s <span>browsing
+   context</span>, <var>W</var>'s <span>browsing context</span>, <var>P</var>, the <span>current
+   global object</span> and the <span>current settings object</span>.</p></li>
```
This passes one more parameter than the algorithm accepts; probably the current global object is unnecessary?

domenic commented on source:
```diff
@@ -80923,10 +80957,28 @@ interface <dfn>BarProp</dfn> {
+   data-x="coop-unsafe-none">unsafe-none</code>".</p></li>
+
+   <li><p>A <dfn data-x="coop-struct-report-endpoint">reporting endpoint</dfn>, which is string or
+   null, initially null.</p></li>
```
I'm wondering why you have null as a possibility, where COEP uses empty string. In particular I notice that COOP reporting bails out for null endpoints, whereas COEP reporting does not. Is this divergence intentional?

domenic replied:
This makes perfect sense; thanks for explaining.

Let's leave the minor null-vs-empty-string inconsistency for now. I like null better, and maybe eventually will update COEP to use it.

domenic commented on source:
```diff
@@ -81112,6 +81388,683 @@ interface <dfn>BarProp</dfn> {
+  <p>To <dfn data-x="coop-check-access-report">check if an access between two browsing contexts
+  should be reported</dfn>, given two <span data-x="browsing context">browsing contexts</span>
+  <var>accessor</var> and <var>accessed</var>, a JavaScript property name
+  <var>P</var>, and an <span>environment</span> <var>environement</var>:</p>
```
I think this should be "an environment settings object", not an environment; we want to use the more narrow type since it's possible.

domenic commented on source:
```diff
@@ -81112,6 +81388,683 @@ interface <dfn>BarProp</dfn> {
+  <var>P</var>, and an <span>environment</span> <var>environement</var>:</p>
+
+  <ol>
+   <li><p>If <var>P</var> is not part of <span>CrossOriginProperties</span>, then return.</p></li>
```
The problem noted at https://github.com/whatwg/html/pull/5518#discussion_r470209008 remains.

domenic commented on source:
```diff
@@ -81112,6 +81388,683 @@ interface <dfn>BarProp</dfn> {
+      </tr>
+      <tr>
+       <td>referrer</td>
+       <td><var>referrer</var></td>
```
This should probably be something like

> _referrer_, [serialized](https://url.spec.whatwg.org/#concept-url-serializer) if it is a URL

or maybe you could create a _referrerString_ variable in earlier steps. This applies in all places that take a _referrer_.

domenic commented on source:
```diff
@@ -81112,6 +81388,683 @@ interface <dfn>BarProp</dfn> {
+      </tr>
+      <tr>
+       <td>type</td>
+       <td><code data-x="">navigation-to-response</code></td>
```
All of the type values should have quotes around them (`"<code data-x="">navigation-to-response</code>"`).

domenic commented on source:
```diff
@@ -83353,17 +84309,25 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+       <li><p>Let <var>coopEnforcementResult</var> be a new  is a new <span
+       data-x="coop-enforcement-result">cross-origin opener policy enforcement result</span> given
+       false, false, <var>resource</var>'s <span data-x="concept-response-url">url</span>,
+       <var>responseOrigin</var>, <var>coop</var> and false.</p></li>
```
This should explicitly state each item's name in the struct, like the next line does for _navigationParams_ and like line 84481 does.

camillelamy replied:
Done.

domenic commented on source:
```diff
@@ -83388,6 +84352,13 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+       false, false, <var>resource</var>'s <span data-x="concept-request-url">url</span>,
+       <var>activeDocumentNavigationOrigin</var>, <var>browsingContext</var>'s <span>active
+       document</span>'s <span data-x="concept-document-coop">cross-origin opener policy</span> and
+       false.</p></li>
```
Same as above.

camillelamy replied:
Done.

domenic reviewed (commented):
I pushed a commit with some typo and nit fixes. The only substantial question left is the use of null for the endpoints, and how that causes divergence (I think in behavior, as well as spec structure) from COEP.

camillelamy commented:
@domenic PTAL. And also sorry, it seems I messed up with the github interface and my previous review comments were not sent.

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+  <span>environment</span> <var>environement</var>:</p>
+
+  <ol>
+   <li><p>If <var>P</var> is not part of <span>CrossOriginProperties</span>, return.</p></li>
```
Yes, I think just nesting it under https://html.spec.whatwg.org/#crossoriginproperties-(-o-) (outside the `<ol>`) would be fine.

domenic commented on source:
```diff
@@ -80457,50 +80664,145 @@ interface <dfn>BarProp</dfn> {
-     <li><p><var>activeDocumentCOOP</var> is "<code
-     data-x="coop-same-origin-allow-popups">same-origin-allow-popups</code>".</p></li>
+     <li>
+      <p>If <var>browsingContext</var>'s <span>browsing context group</span>'s <span>browsing
```
Ah OK. In that case this should change to `<var>browsingContext</var>'s <span data-x="tlbc group">group</span>`

domenic commented on source:
```diff
@@ -80527,6 +80829,623 @@ interface <dfn>BarProp</dfn> {
+     document</span>'s <span>origin</span>, <var>P</var>, <var>accessor</var>'s <span>top-level
+     browsing context</span>'s <span>active document</span>'s <span
+     data-x="dom-document-referrer">referrer</span>, <var>globalObject</var>, and
+     <var>environment</var>.</p></li>
```
The new algorithm LGTM, and indeed it's nice that it centralizes the pulling out of properties of the browsing context into one place. It's still a *lot* of parameters, but they're nicely centralized, and hiding those parameters into the browsing-context god object would probably make things worse.

domenic commented on source:
```diff
@@ -84312,10 +84327,17 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+       data-x="coop-enforcement-result">cross-origin opener policy enforcement result</span> whose
+       <span data-x="coop-enforcement-bcg-switch">needs a browsing context group switch</span> is
+       false, <span data-x="coop-enforcement-bcg-switch-report-only">would need a browsing context
+       group switch due to report-only</span> is false,<span
```
Missing space

camillelamy replied:
Done.

domenic commented on source:
```diff
@@ -84353,11 +84375,18 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+       data-x="coop-enforcement-result">cross-origin opener policy enforcement result</span> whose
+       <span data-x="coop-enforcement-bcg-switch">needs a browsing context group switch</span> is
+       false, <span data-x="coop-enforcement-bcg-switch-report-only">would need a browsing context
+       group switch due to report-only</span> is false,<span
```
Missing space

camillelamy replied:
Done.

domenic reviewed (approved):
Overall LGTM. What remains are the following nits (denoted above by unresolved comments, but centralized here in case it's helpful): two missing spaces, improving the precision for cross-origin accessible window property name, and changing the link for "group".

I believe there's been good progress on adding web platform tests for this feature as well, right? Can you update the OP to link to them, and comment on how complete you think the coverage is?

Finally, for multi-implementer interest, perhaps @annevk can comment for Firefox?

annevk commented:
I think I said before that Firefox is supportive of reporting for COOP and COEP, though it's not a priority for us. (I also worry about the complexity this adds due to the unusual requirements of COOP reporting, but hopefully adequate test coverage, in particular for all the various origin checks, can mitigate some of that.)

camillelamy commented:
Thanks Domenic! I should have fixed all remaining issues in the latest update, and updated the pull request. In terms of WPTs, I believe we have thorough testing for all base cases for access reporting, including the URLs to send in reports. For browsing context switch, we're at least missing tests for 2 base cases that I am trying to add right now. After that I think we should be good. We have generally tested each interesting part of the feature on its own, and haven't done combinatorial tests.

domenic commented:
Awesome! I'll merge this momentarily, then. The test coverage sounds like it's definitely enough to merge the spec change, although of course more is always better.

sideshowbarker commented:
cc @whatwg/documentation
